tools/testharness.py

```diff
@@ -1,547 +1,535 @@
-                    break
-            for e, lines in exceptions:
-                tc=TestCase(e[1].name,
-                            '%s.%s'%(e[1].length,
-                                     e[1].filename[:-4]))
-                tc.add_failure_info("Failure", lines)
-                self.xml_parser.test_cases+= [tc]
-                self.tests.remove(e)
-                self.completed_tests += [e[1]]
-
-            count = len(self.tests)
-            while True:
-                for t in self.tests:
-                  if t is None: continue
-                  test = t[1]
-                  os.chdir(t[0])
-                  if test.is_finished():
-                      if test.length == "long":
-                        test.fl_logs(nLogLines = 20)
-                      else:
-                        test.fl_logs(nLogLines = 0)
-                      try:
-                        self.teststatus += test.test()
-                      except:
-                        self.log("Error: %s raised an exception while testing:" % test.filename)
-                        lines = traceback.format_exception( sys.exc_info()[0], sys.exc_info()[1], sys.exc_info()[2] )
-                        for line in lines:
-                            self.log(line)
-                        self.teststatus += ['F']
-                        test.pass_status = ['F']
-                      self.completed_tests += [test]
-                      self.xml_parser.test_cases+=test.xml_reports
-                      t = None
-                      count -= 1
-
-                if count == 0: break
-                time.sleep(60)                  
-        else:
-          for t in self.tests:
-            test = t[1]
-            os.chdir(t[0])
-            if self.length == "long":
-              test.fl_logs(nLogLines = 20)
+        # Define conditions the test must meet
+        # Do we want to keep this condition, or do we want to have "any"
+        # correspond to all tests when length is not given (and only run the
+        # provided lengths when length is given)?
+        length_condition = (
+            (args.length == "any" and prob_length in ["long", "vlong"])
+            or (args.length != "any" and prob_length not in args.length))
+        nprocs_condition = (
+            (args.parallel == "parallel" and prob_nprocs <= 1)
+            or (args.parallel == "serial" and prob_nprocs != 1))
+        excluded_tags_condition = set(xml_tags).intersection(args.exclude_tags)
+        required_tags_condition = set(args.tags).difference(xml_tags)
+        # Skip the test if any of the conditions is not met
+        if (length_condition or nprocs_condition or excluded_tags_condition
+                or required_tags_condition):
+            xml_entry = TestCase(
+                name=xml_file.stem, classname=prob_classname, status="Skipped")
+            xml_entry.add_skipped_info(message="Test suite conditions unmet.")
+            xml_parser.test_cases.append(xml_entry)
```

Comment:
Hm, this is a bit wonky - you're modifying a global variable, which is unhelpfully called `xml_parser` whereas it seems to be a `TestSuite`. So `test_suite` or `junit_test_suite` would be a better name? And `test_case` instead of `xml_entry`? (I realize this was bad in the old testharness already)
But more importantly could you just return a list of skipped test cases? Or add it in the same dictionary but with a flag `"skipped": True` and create the junit stuff all in one place at the end...

Reply:
Agreed with the naming conventions; is it a problem that `test_suite` is defined in the outermost scope? 
Regarding the skipped tests, they should already be included in the XML report. What did you have in mind with that list idea?
The problem with creating all TestCase entries for the JUnit report at the end is that different arguments are involved in the call depending on the test's final status and so multiple calls to TestCase arranged in an `if, elif, else` block would still be required. I am not sure this is a better way to do it.

Reply:
It's just bad code structure. As a general rule you should only use global variables (and even more so modify them) in exceptional cases. Otherwise your code will be very hard to understand/change later on. When I'm reading that routine I have no idea where test_suite is being defined. When I'm actually reading the bottom of the code where test_suite is created and later on the completed test results are appended, I have no way of telling that the inbetween call to `filter_tests` also modifies `test_suite`. So it becomes very hard to reason about the data flow - and for instance I might be tempted to move the creation of the `test_suite` to after the call to `filter_tests`.
So as a minimum you should have test_suite as an argument to filter_tests - to signify that this routine is doing something to that test_suite, and then tell us at the top what it is: namely "This routine filters tests that fulfill our chosen criteria and are returned in a dictionary. Information about skipped tests is stored in the junit TestSuite". Which brings us to the second part of what I suggested (but that's more of a suggestion and if it doesn't work easily feel free to ignore) - if you think about your routine in this way it becomes clear that it is doing multiple things to multiple data structures at the same time. Again as a general rule, you want to break up your code in such a way that each function has a single responsibility. This makes it much easier to later on to modify the code. Even when you're not using classes, try to think about your routines in terms of the data structures it operates on and ideally only modify one structure at the time.